core/src/main/java/hudson/diagnosis/OldDataMonitor.java

```diff
@@ -218,19 +216,28 @@ public static void report(Saveable obj, Collection<Throwable> errors) {
             return;
         }
         OldDataMonitor odm = get(j);
-        synchronized (odm) {
-            SaveableReference ref = referTo(obj);
+        SaveableReference ref = referTo(obj);
+        while (true) {
             VersionRange vr = odm.data.get(ref);
-            if (vr != null) vr.extra = buf.toString();
-            else            odm.data.put(ref, new VersionRange(null, buf.toString()));
+            if (vr != null) {
+                vr.extra = buf.toString();
+                break;
+            } else if (odm.data.put(ref, new VersionRange(null, buf.toString())) == null) {
```

Comment:
putIfAbsent? Otherwise there is a risk of overriding the VersionRange if it gets inserted by another thread